File: day-03-1/ruby/answer.rb
# frozen_string_literal: true

class Answer
  def self.read_input_file
    File.read('input.txt')
  end

  def count_ones(data, index)
    data.reduce(0) { |count, number| count + number[index].to_i }
  end

  def most_common(data, index)
    count_ones(data, index) > data.size / 2.0 ? 1 : 0
  end

  def main
    diagnostic_report = self.class.read_input_file.split("\n")
    number_size = diagnostic_report.first.size

    gamma_rate = (0...number_size).reduce(0) do |rate, index|
      (rate << 1) + most_common(diagnostic_report, index)
    end
    epsilon_rate = gamma_rate ^ (2**number_size - 1)

    power_consumption = gamma_rate * epsilon_rate
    power_consumption
  end
end

puts Answer.new.main

File: day-03-1/java/src/main/java/com/adventofcode/Answer.java
package com.adventofcode;

import java.io.FileInputStream;
import java.io.IOException;

class Answer {

    public static String readInputFile() {
        StringBuilder strBuilder = new StringBuilder();
        int readVar;

        try (FileInputStream fileInput = new FileInputStream("input.txt")) {
            do {
                readVar = fileInput.read();
                if (readVar != -1) {
                    strBuilder.append((char) readVar);
                }
            } while (readVar != -1);
        } catch (IOException e) {
            System.out.println("I/O Error: " + e);
            return "";
        }

        return strBuilder.toString();
    }

    public int solution() {
        String[] diagnosticReport = readInputFile().split("\n");
        int numberSize = diagnosticReport[0].length();
        int gammaRate = 0;
        int epsilonRate;
        int powerConsumption;

        for (int i = 0; i < numberSize; i++) {
            int indexOneCount = 0;

            for (String binNumber : diagnosticReport) {
                indexOneCount += binNumber.charAt(i) - '0';
            }

            gammaRate <<= 1;
            if (indexOneCount > diagnosticReport.length / 2.0) {
                gammaRate++;
            }
        }

        epsilonRate = gammaRate ^ ((int) Math.pow(2, numberSize) - 1);
        powerConsumption = gammaRate * epsilonRate;
        return powerConsumption;
    }

    public static void main(String[] args) {
        System.out.println(new Answer().solution());
    }

}
